### ISSUE TITLE
acquireTokenSilent after first redirect is not working

### ISSUE THREAD

**Issue by jfbloom22** [AzureAD/microsoft-authentication-library-for-js#1219]
## Library
- [x] `msal@1.2.1`

## Framework
React 16.5.2

### With 3rd Party cookies enabled
After a redirect login, the app immediately renews the token in a hidden iFrame.

### With 3rd Party cookies disabled
After a redirect login, the app tries to renew the token in a hidden iFrame, this fails, then it redirects again (it does not prompt the user again, it just disrupts the login flow with an extra redirect)

## Security
No

## Regression
This behavior has been the same since 1.0.2.

## Configuration
Please provide your MSAL configuration options.

```js
const MSAL_AUTHORITY = `https://login.microsoftonline.com/tfp/${
  process.env.REACT_APP_MSAL_TENANT_ID
}/${process.env.REACT_APP_MSAL_SIGNIN_POLICY}/v2.0`;
export const MSAL_SCOPES = {
  MMG: `https://${process.env.REACT_APP_MSAL_TENANT_NAME}.onmicrosoft.com/${
    process.env.REACT_APP_MSAL_APP_ID_URI
  }/read`,
  OPENID: 'openid'
};
export const msalApp = new UserAgentApplication({
  auth: {
    clientId: `${process.env.REACT_APP_MSAL_CLIENT_ID}`,
    authority: MSAL_AUTHORITY,
    validateAuthority: true,
    postLogoutRedirectUri: `${process.env.REACT_APP_HOST_DOMAIN}`,
    redirectUri: `${process.env.REACT_APP_HOST_DOMAIN}`,
    navigateToLoginRequestUrl: false
  },
  cache: {
    cacheLocation: 'sessionStorage',
    storeAuthStateInCookie: false
  },
  system: {
    // increase the number of milliseoconds of inactivity before a token renewal response from Azure AD should be considered timed out.
    loadFrameTimeout: 9000,
    tokenRenewalOffsetSeconds: 0 
  }
});


export const acquireToken = () => {
  const request: AuthenticationParameters = {
    scopes: [MSAL_SCOPES.MMG]
  };
  console.log('MSAL account from inside acquireToken() ', msalApp.getAccount())
  return msalApp.acquireTokenSilent(request).catch(error => {
    // Call acquireTokenPopup (popup window) in case of acquireTokenSilent failure
    // due to consent or interaction required ONLY
    console.log('catching error acquireTokenSilent', error);
    if (requiresInteraction(error.errorCode)) {
      console.error('redirecting to login from inside acquireTokenSilent');
      msalApp.acquireTokenRedirect(request)
      return delay(3000).then(() => {
        // wait for the redirect before returning the error to the calling function.
        console.log('redirecting to MSALlogin took too long, throwing error.')
        throw error;
      });
    }
    throw error;
  });
};
```

These are for a test app, not production, so I can share:
```
REACT_APP_MSAL_CLIENT_ID=6fa0ff31-bfad-4d6a-99d2-3f7d00cc0b2a
REACT_APP_MSAL_TENANT_ID=7bec88f8-013e-403f-8cc9-9e451c46f7ea
REACT_APP_MSAL_SIGNIN_POLICY=B2C_1_MMG_Test_Signin
REACT_APP_MSAL_FORGET_PASSWORD_POLICY=B2C_1_MMG_ForgotPassword_Test
REACT_APP_MSAL_TENANT_NAME=InteractionPlacesB2C1
REACT_APP_MSAL_APP_ID_URI=mmg-api
```
### With 3rd party cookies enabled
1. acquireTokenRedirect()
2. acquireTokenSilent()
3. observe that acquireTokenSilent renews the session in a hidden iFrame

How do I know it is loading in the iFrame?
Because I have a little snippet of code that prevents my Javascript from running inside the iFrame:
```
if (!(window !== window.parent && !window.opener)) {
// my app initialization code goes here
} else {
  console.log('loading silent refresh in iFrame');
}
```

### 3rd party cookies disabled
1. acquireTokenRedirect()
2. acquireTokenSilent()
3. observe that acquireTokenSilent tries to the session in a hidden iFrame
4. iFrame renewal fails and acquireTokenRedirect() is called
5. App redirects (it does not prompt the user again, it just disrupts the login flow with an extra redirect)
6. acquireTokenSilent() returns the accessToken

## Expected behavior
After acquireTokenRedirect() returns, acquireTokenSilent() should return the cached accessToken rather than trying to renew the session in a hidden iFrame.

FYI this line:
`console.log('MSAL account from inside acquireToken() ', msalApp.getAccount())`
returns the expected account. 

## Browsers
lates Chrome for Mac and Safari

**Comment by jasonnutter** [comment]
@jfbloom22 It is a known issue that MSAL is unable to silently acquire an access token if third-party cookies are disabled. This is because when we attempt to acquire a token through a hidden iframe, that iframe needs to have access to your AAD session, and in that context, the cookies for your AAD session for your authority domain are considered third-party (relative to your application domain), so AAD doesn't think you have an active session, which results in an error in MSAL.

We are currently working on the next version of the library, which will switch to the Auth Code Flow w/ PKCE, which will mitigate the problems with third-party cookies in most scenarios.

**Comment by jasonnutter** [comment]
Oh, nevermind, I think I misunderstood your issue.

I wonder if you have a race condition, where you are attempting to read the tokens from the cache before the response has been processed? Try inspecting local storage right before you call `acquireTokenSilent`. 

The `handleRedirectCallback` will also give you the redirect response, which will include the access token.

**Comment by jfbloom22** [comment]
Thanks @jasonnutter.  I am very excited about the `Auth Code Flow w/ PKCE` coming down the pipe.  Can't wait!  

Here is session storage right after logging in and before it calls acquireTokenSilent():  https://www.dropbox.com/s/jqrgkydm1oaqpf4/Screenshot%202020-01-16%2017.14.09.png?dl=0

**Comment by jasonnutter** [comment]
@jfbloom22 Take a look at the composite key for the access token, and make sure that your request for `acquireTokenSilent` matches the values of that composite key for scopes, authority, etc.

**Comment by jfbloom22** [comment]
Is there another way I can verify the composite keys?  I double checked my code, and I am clearly passing in the same scopes.  The authority is set in only one spot, when `msalApp` is defined.

Initial login:  
``` 
 const request: AuthenticationParameters = {
      scopes: [MSAL_SCOPES.MMG]
    };
    return msalApp.loginRedirect(request);
```

then `acquireTokenSilent`
```
const request: AuthenticationParameters = {
    scopes: [MSAL_SCOPES.MMG]
  };
  return msalApp.acquireTokenSilent(request).catch(error => {
```

One curios bit is that the key for the value with the `accessToken`.  Here is the key after initial login:
```
{
    "authority": "https://login.microsoftonline.com/tfp/7bec88f8-013e-403f-8cc9-9e451c46f7ea/b2c_1_mmg_test_signin/",
    "clientId": "6fa0ff31-bfad-4d6a-99d2-3f7d00cc0b2a",
    "scopes": "6fa0ff31-bfad-4d6a-99d2-3f7d00cc0b2a",
    "homeAccountIdentifier": "YTJhZGFmZjctYjAzNi00ODY0LWI0ZDQtN2NjM2MzNTA3MWUzLWIyY18xX21tZ190ZXN0X3NpZ25pbg==.N2JlYzg4ZjgtMDEzZS00MDNmLThjYzktOWU0NTFjNDZmN2Vh"
}
```
Then after `acquireTokenSilent` there is an additional key:

```
{
        "authority": "https://login.microsoftonline.com/tfp/7bec88f8-013e-403f-8cc9-9e451c46f7ea/b2c_1_mmg_test_signin/",
        "clientId": "6fa0ff31-bfad-4d6a-99d2-3f7d00cc0b2a",
        "scopes": "https://InteractionPlacesB2C1.onmicrosoft.com/mmg-api/read",
        "homeAccountIdentifier": "YTJhZGFmZjctYjAzNi00ODY0LWI0ZDQtN2NjM2MzNTA3MWUzLWIyY18xX21tZ190ZXN0X3NpZ25pbg==.N2JlYzg4ZjgtMDEzZS00MDNmLThjYzktOWU0NTFjNDZmN2Vh"
    }
```
Notice that the `scopes` are different.  After the initial login the `scopes` include only the clientID.  Should I expect this `scopes` value to be the value I passed in?  When I found this, I though surely I am calling login with the clientID as the only scope.  But I have checked this like 5 times and I am definitely passing in `"https://InteractionPlacesB2C1.onmicrosoft.com/mmg-api/read"` for both `loginRedirect` and `acquireTokenSilent`.  Is it possible I am formatting the `request` object incorrectly for `loginRedirect`?

Also, I am very curious if anyone else is seeing this behavior.  Should be easy enough to replicate, configure a redirect style login, then call `acquireTokenSilent` right after logging in.  Include some console logging like I have above to log when the silent iFrame loads.  Set your dev tools to preserve the logs and observe.

**Comment by jasonnutter** [comment]
@jfbloom22 Client Id as the scope would indicate an ID token, which would make sense after a login call. Does this happen after `acquireTokenRedirect` as well, or only `loginRedirect`?

**Comment by jasonnutter** [comment]
@jfbloom22 That means MSAL doesn't think the user is logged in. You should check if `msalApp.getAccount()` is truthy, and if not, invoke one of the login methods before calling one of the acquireToken methods.

**Comment by tnorling** [comment]
@jfbloom22 Is this still an issue?

**Comment by tnorling** [comment]
We believe the original issue was fixed in `msal@1.3.0`. Let us know if that's not the case. Thanks!
